clippy_lints/src/doc.rs

```diff
@@ -34,6 +34,39 @@ declare_clippy_lint! {
     "presence of `_`, `::` or camel-case outside backticks in documentation"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for the doc comments of publicly visible
+    /// unsafe functions and warns if there is no `# Safety` section.
+    ///
+    /// **Why is this bad?** Unsafe functions should document their safety
+    /// preconditions, so that users can be sure they are using them safely.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Examples**:
+    /// ```rust
+    /// /// This function should really be documented
+    /// unsafe fn start_apocalypse(u: &mut Universe) {
+    ///     unimplemented!();
+    /// }
+    /// ```
+    ///
+    /// At least write a line about safety:
+    ///
+    /// ```rust
+    /// /// # Safety
+    /// ///
+    /// /// This function should not be called before the horsemen
+    /// /// are ready.
+    /// unsafe fn start_apocalypse(u: &mut Universe) {
+    ///     unimplemented!();
+    /// }
+    /// ```
+    pub MISSING_SAFETY_DOC,
+    correctness,
```

Comment:
I don't think this should be a deny-by-default lint, since it doesn't break the code, if it is missing. Maybe a `style` lint, since it is specified in the API guidelines, that this should be done.